### PR TITLE
[id-split] part 1: feature extraction

### PR DESCRIPTION
**Pull request by EgorBu**
* feature extraction
* tests
* misc for travis and setup

### CODE REVIEW

File: sourced/ml/algorithms/id_splitter/features.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+# CSV default parameters
+TOKEN_COL = 3
+TOKEN_SPLIT_COL = 4
+
+
+def read_identifiers(csv_loc, use_header=True, mode="r", maxlen=MAXLEN, token_col=TOKEN_COL,
+                     token_split_col=TOKEN_SPLIT_COL, shuffle=True):
+    """
+    Read and filter too long identifiers from CSV file.
+    :param csv_loc: location of CSV file
```
Comment (r0mainK) [inline sourced/ml/algorithms/id_splitter/features.py]: You say CSV file but it's actually tared CSV file, think you should either update dovs/log or check path to see whether it is in fact a tar file or csv file that was inputted

File: sourced/ml/algorithms/id_splitter/features.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    log.info("Reading data from CSV...")
+    identifiers = []
+    with tarfile.open(csv_loc, mode=mode, encoding="utf-8") as f:
+        assert len(f.members) == 1, "Expect one archived file"
+        content = f.extractfile(f.members[0])
+        if not use_header:
+            content.readline()
+        for line in content:
+            parts = line.decode("utf-8").strip().split(",")
+            if len(parts[token_col]) <= maxlen:
```
Comment (r0mainK) [inline sourced/ml/algorithms/id_splitter/features.py]: Instead of setting the max length why not add it as an argument which defaults to 40 ?

File: sourced/ml/algorithms/id_splitter/features.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                split_arr.append(1)
+                skip_char = True
+        # sanity check
+        assert len(index_arr) == len(split_arr)
+        char_id_seq.append(index_arr)
+        splits.append(split_arr)
+
+    # train/test splitting
+    log.info("Train/test splitting...")
+    n_train = int((1 - test_size) * len(char_id_seq))
```
Comment (r0mainK) [inline sourced/ml/algorithms/id_splitter/features.py]: I might be mistaken but I could not find in any of the PRs another reference to test_size not being used this way: `1 - test_size`. You could directly use train_size fraction instead ?

File: sourced/ml/algorithms/id_splitter/features.py (hunk 1 above)
Comment (EgorBu) [reply sourced/ml/algorithms/id_splitter/features.py]: yes, this part should be updated according to @zurk comments because he releases dataset.

File: sourced/ml/algorithms/id_splitter/features.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    x_tr = char_id_seq[:n_train]
+    x_t = char_id_seq[n_train:]
+    y_tr = splits[:n_train]
+    y_t = splits[n_train:]
+    log.info("Number of train samples: {}, number of test samples: {}.".format(len(x_tr),
```
Comment (vmarkovtsev) [inline sourced/ml/algorithms/id_splitter/features.py]: `log.info("Number of train samples: %s, number of test samples: %s.", len(x_tr), len(x_t))`

File: sourced/ml/algorithms/id_splitter/features.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        split_arr = []
+        skip_char = False
+        for char in identifier.strip():
+            if char in char2ind:
+                index_arr.append(char2ind[char])
+                if skip_char:
+                    skip_char = False
+                    continue
+                split_arr.append(0)
+            else:
```
Comment (zurk) [inline sourced/ml/algorithms/id_splitter/features.py]: In case of bad data better to have 
`elif char == " "` here 
and in the end
```python
else:
    log.warning("Unexpected symbol %s in identifier", char)
```

File: sourced/ml/algorithms/id_splitter/features.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        # sanity check
+        assert len(index_arr) == len(split_arr)
+        char_id_seq.append(index_arr)
+        splits.append(split_arr)
+
+    # train/test splitting
+    log.info("Train/test splitting...")
+    n_train = int((1 - test_size) * len(char_id_seq))
+    x_tr = char_id_seq[:n_train]
+    x_t = char_id_seq[n_train:]
```
Comment (zurk) [inline sourced/ml/algorithms/id_splitter/features.py]: `x_tr ` and `x_t` are hard to distinguish. 
let's use full names:
`x_train` and `x_test`

File: sourced/ml/algorithms/id_splitter/features.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    """
+    log = logging.getLogger("id-splitter-prep")
+    log.info("Reading data from CSV...")
+    identifiers = []
+    with tarfile.open(csv_loc, mode=mode, encoding="utf-8") as f:
+        assert len(f.members) == 1, "Expect one archived file"
+        content = f.extractfile(f.members[0])
+        if not use_header:
+            content.readline()
+        for line in content:
```
[the remark this answers is not in the document]
Comment (warenlg) [reply sourced/ml/algorithms/id_splitter/features.py]: @k for the progress bar, what used to be done in ml ? I mean I don't want to import a package just for this like `clint` ?
Or do you prefer something like this ? 
```for i, line in enumerate(content):
            if i % 1000 == 0:
    sys.stderr.write("%d\r" % i)
```

[the remark this answers is not in the document]
Comment (vmarkovtsev) [reply sourced/ml/algorithms/id_splitter/features.py]: E.g. https://github.com/src-d/ml/pull/254/files/e2fbeb746792c9e235f67e49c5e04ff89a4f5b6e..e7b997518dfdf764e4d8e1bcc42e7b61760cee50#diff-f61cfa96b08f3a41b6cc8a3e1afebd6fL5

[the remark this answers is not in the document]
Comment (zurk) [reply sourced/ml/algorithms/id_splitter/features.py]: right, we have it in modelforge just use `progress_bar`
`from modelforge.progress_bar import progress_bar`

File: sourced/ml/algorithms/id_splitter/features.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+                     split_identifiers_col: int=CSV_SPLIT_IDENTIFIERS_COL,
+                     shuffle: bool=True):
+    """
+    Reads and filters too long identifiers from CSV file.
+    :param csv_path: path to the CSV file.
+    :param use_header: uses header as normal line (True) or treat as header line with column names.
+    :param max_identifier_len: maximum length of raw identifier. Skip identifier if longer.
+    :param identifiers_col: column in CSV file for the raw identifier.
+    :param split_identifiers_col: column in CSV file for the splitted identifier.
+    :param shuffle: indicates whether to reorder the list of identifiers at random after reading it
```
Comment (EgorBu) [inline sourced/ml/algorithms/id_splitter/features.py]: missed dot at the end

File: sourced/ml/tests/test_id_splitter_features.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    # generate file
+    res = []
+    for sz in char_sizes:
+        line = ",".join([text * sz] * n_cols)
+        res.append(line)
+    content = "\n".join(res)
+    content = content.encode("utf-8")
+
+    # add content to file
+    info = tarfile.TarInfo('identifiers.txt')
```
Comment (EgorBu) [inline sourced/ml/tests/test_id_splitter_features.py]: wrong quotes

File: sourced/ml/models/tensorflow_model.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    @property
+    def outputs(self):
+        """
+        Returns the list of outputs of the tensorflow graph.
+        """
+        return self._outputs
+
+    def serialize(self):
+        """
+        Exports the Protobuf tensorflow model into a GraphDef object
```
Comment (EgorBu) [inline sourced/ml/models/tensorflow_model.py]: missed dots

File: sourced/ml/models/tensorflow_model.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
+from typing import List
```
Comment (EgorBu) [inline sourced/ml/models/tensorflow_model.py]: this file is not from this PR

File: sourced/ml/tests/test_id_splitter_features.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+import numpy as np
+
+from sourced.ml.algorithms.id_splitter.features import prepare_features, read_identifiers
+from sourced.ml.tests.models import IDENTIFIERS
+
+
+def write_fake_identifiers(tar_file, n_lines, char_sizes, n_cols, text="a"):
+    """
+    Prepare file with fake identifiers.
+    :param tar_file: ready to write file
```
Comment (EgorBu) [inline sourced/ml/tests/test_id_splitter_features.py]: missed dots in the end

File: .travis.yml [hunk 13]
```diff
@@ -15,7 +15,7 @@ _install: &_install
   - gimme 1.8
   - source ~/.gimme/envs/latest.env
   - pip install --upgrade pip
-  - pip install tensorflow
+  - pip install tensorflow keras
   - pip install -r requirements.txt codecov
```
Comment (vmarkovtsev) [inline .travis.yml]: We decided to put Keras into the requirements because it does not depend on Tensorflow directly. Thus this is not needed

File: setup.py [hunk 14]
```diff
@@ -37,6 +37,7 @@
     extras_require={
         "tf": ["tensorflow>=1.0,<2.0"],
         "tf_gpu": ["tensorflow-gpu>=1.0,<2.0"],
+        "keras": ["keras>=2.0,<3.0"],
```
Comment (vmarkovtsev) [inline setup.py]: Can be safely added to the main requirements - it does not depend on Tensorflow

File: sourced/ml/algorithms/id_splitter/features.py [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+    :param use_header: uses header as normal line (True) or treat as header line with column names.
+    :param max_identifier_len: maximum length of raw identifier. Skip identifier if longer.
+    :param identifiers_col: column in CSV file for the raw identifier.
+    :param split_identifiers_col: column in CSV file for the splitted identifier.
+    :param shuffle: indicates whether to reorder the list of identifiers
+        at random after reading it.
+    :return: list of splitted identifiers.
+    """
+    log = logging.getLogger("read_identifiers")
+    log.info("Reading data from CSV...")
```
Comment (vmarkovtsev) [inline sourced/ml/algorithms/id_splitter/features.py]: Please include the CSV path to the log

File: sourced/ml/algorithms/id_splitter/features.py [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+    identifiers = []
+    # TODO: Update dataset loading as soon as https://github.com/src-d/backlog/issues/1212 done.
+    # Think about dataset download step.
+    with tarfile.open(csv_path, mode="r", encoding="utf-8") as f:
```
Comment (vmarkovtsev) [inline sourced/ml/algorithms/id_splitter/features.py]: `mode="r"` is not needed

File: sourced/ml/algorithms/id_splitter/features.py [hunk 17]
```diff
@@ -0,0 +1,125 @@
+import logging
+import string
+import tarfile
+import warnings
+
+import numpy as np
```
Comment (vmarkovtsev) [inline sourced/ml/algorithms/id_splitter/features.py]: Can you please check how do we normally import numpy in this project. If it is `numpy` in most of the places, should be changed

File: sourced/ml/algorithms/id_splitter/features.py [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+        np.random.shuffle(identifiers)
+    log.info("Number of identifiers after filtering: %s." % len(identifiers))
+    return identifiers
+
+
+def prepare_features(csv_path: str, use_header: bool=True,
+                     max_identifier_len: int=DEFAULT_MAX_IDENTIFIER_LEN,
+                     identifiers_col: int=CSV_IDENTIFIERS_COL,
+                     split_identifiers_col: int=CSV_SPLIT_IDENTIFIERS_COL,
+                     shuffle: bool=True, test_size: float=0.2, padding: str=PADDING):
```
Comment (vmarkovtsev) [inline sourced/ml/algorithms/id_splitter/features.py]: `0.2` should be moved to global constants

Default shuffle value `True` too

File: sourced/ml/algorithms/id_splitter/features.py [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+    log = logging.getLogger("prepare_features")
+
+    # read data from file
+    identifiers = read_identifiers(csv_path=csv_path, use_header=use_header,
+                                   max_identifier_len=max_identifier_len,
+                                   identifiers_col=identifiers_col,
+                                   split_identifiers_col=split_identifiers_col, shuffle=shuffle)
+
+    # convert identifiers into character indices and labels
+    log.info("Converting identifiers to character indices...")
```
Comment (vmarkovtsev) [inline sourced/ml/algorithms/id_splitter/features.py]: Can you please add the number of identifiers and probably other useful stats to this log

File: sourced/ml/algorithms/id_splitter/features.py (hunk 8 above)
Comment (warenlg) [reply sourced/ml/algorithms/id_splitter/features.py]: This was adressed even though the diff is not showing it
